{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "alive-century",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "decent-warning",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2785510708.tar.gz\n",
      "2785510708.cog.tab.txt\n",
      "2785510708.fna\n",
      "2785510708.genes.faa\n",
      "2785510708.genes.fna\n",
      "2785510708.gff\n",
      "2785510708.intergenic.fna\n",
      "2785510708.ko.tab.txt\n",
      "2785510708.ko.tab.txt\n",
      "2785510708.pfam.tab.txt\n",
      "2785510708.pfam.tab.txt\n",
      "2785510708.signalp.tab.txt\n",
      "2785510708.signalp.tab.txt\n",
      "2785510708.tigrfam.tab.txt\n",
      "2785510708.tigrfam.tab.txt\n",
      "2785510708.tmhmm.tab.txt\n",
      "2785510708.tmhmm.tab.txt\n",
      "README.txt\n",
      ".ipynb_checkpoints\n",
      "2814122942.cog.tab.txt\n",
      "2814122942.crispr.txt\n",
      "2814122942.fna\n",
      "2814122942.genes.faa\n",
      "2814122942.genes.fna\n",
      "2814122942.gff\n",
      "2814122942.intergenic.fna\n",
      "2814122942.ko.tab.txt\n",
      "2814122942.ko.tab.txt\n",
      "2814122942.pfam.tab.txt\n",
      "2814122942.pfam.tab.txt\n",
      "2814122942.signalp.tab.txt\n",
      "2814122942.signalp.tab.txt\n",
      "2814122942.tigrfam.tab.txt\n",
      "2814122942.tigrfam.tab.txt\n",
      "2814122942.tmhmm.tab.txt\n",
      "2814122942.tmhmm.tab.txt\n",
      "README.txt\n",
      "2863075697.cog.tab.txt\n",
      "2863075697.crispr.txt\n",
      "2863075697.fna\n",
      "2863075697.genes.faa\n",
      "2863075697.genes.fna\n",
      "2863075697.gff\n",
      "2863075697.intergenic.fna\n",
      "2863075697.ko.tab.txt\n",
      "2863075697.ko.tab.txt\n",
      "2863075697.pfam.tab.txt\n",
      "2863075697.pfam.tab.txt\n",
      "2863075697.signalp.tab.txt\n",
      "2863075697.signalp.tab.txt\n",
      "2863075697.smart.tab.txt\n",
      "2863075697.superfam.tab.txt\n",
      "2863075697.tigrfam.tab.txt\n",
      "2863075697.tigrfam.tab.txt\n",
      "2863075697.tmhmm.tab.txt\n",
      "2863075697.tmhmm.tab.txt\n",
      "README.txt\n",
      "2773857713.tar.gz\n",
      "2773857713.cog.tab.txt\n",
      "2773857713.fna\n",
      "2773857713.genes.faa\n",
      "2773857713.genes.fna\n",
      "2773857713.gff\n",
      "2773857713.intergenic.fna\n",
      "2773857713.ko.tab.txt\n",
      "2773857713.ko.tab.txt\n",
      "2773857713.pfam.tab.txt\n",
      "2773857713.pfam.tab.txt\n",
      "2773857713.signalp.tab.txt\n",
      "2773857713.signalp.tab.txt\n",
      "2773857713.tigrfam.tab.txt\n",
      "2773857713.tigrfam.tab.txt\n",
      "2773857713.tmhmm.tab.txt\n",
      "2773857713.tmhmm.tab.txt\n",
      "README.txt\n",
      "2765236464.tar.gz\n",
      "2765236464.cog.tab.txt\n",
      "2765236464.fna\n",
      "2765236464.genes.faa\n",
      "2765236464.genes.fna\n",
      "2765236464.gff\n",
      "2765236464.intergenic.fna\n",
      "2765236464.ipr.tab.txt\n",
      "2765236464.ko.tab.txt\n",
      "2765236464.ko.tab.txt\n",
      "2765236464.pfam.tab.txt\n",
      "2765236464.pfam.tab.txt\n",
      "2765236464.signalp.tab.txt\n",
      "2765236464.signalp.tab.txt\n",
      "2765236464.tigrfam.tab.txt\n",
      "2765236464.tigrfam.tab.txt\n",
      "2765236464.tmhmm.tab.txt\n",
      "2765236464.tmhmm.tab.txt\n",
      "README.txt\n",
      "2781126040.tar.gz\n",
      "2781126040.cog.tab.txt\n",
      "2781126040.fna\n",
      "2781126040.genes.faa\n",
      "2781126040.genes.fna\n",
      "2781126040.gff\n",
      "2781126040.intergenic.fna\n",
      "2781126040.ko.tab.txt\n",
      "2781126040.ko.tab.txt\n",
      "2781126040.pfam.tab.txt\n",
      "2781126040.pfam.tab.txt\n",
      "2781126040.signalp.tab.txt\n",
      "2781126040.signalp.tab.txt\n",
      "2781126040.tigrfam.tab.txt\n",
      "2781126040.tigrfam.tab.txt\n",
      "2781126040.tmhmm.tab.txt\n",
      "2781126040.tmhmm.tab.txt\n",
      "README.txt\n",
      "2758568662.tar.gz\n",
      "2758568662.cog.tab.txt\n",
      "2758568662.fna\n",
      "2758568662.genes.faa\n",
      "2758568662.genes.fna\n",
      "2758568662.gff\n",
      "2758568662.intergenic.fna\n",
      "2758568662.ipr.tab.txt\n",
      "2758568662.ko.tab.txt\n",
      "2758568662.ko.tab.txt\n",
      "2758568662.pfam.tab.txt\n",
      "2758568662.pfam.tab.txt\n",
      "2758568662.signalp.tab.txt\n",
      "2758568662.signalp.tab.txt\n",
      "2758568662.tigrfam.tab.txt\n",
      "2758568662.tigrfam.tab.txt\n",
      "2758568662.tmhmm.tab.txt\n",
      "2758568662.tmhmm.tab.txt\n",
      "README.txt\n",
      "2765236465.tar.gz\n",
      "2765236465.cog.tab.txt\n",
      "2765236465.fna\n",
      "2765236465.genes.faa\n",
      "2765236465.genes.fna\n",
      "2765236465.gff\n",
      "2765236465.intergenic.fna\n",
      "2765236465.ipr.tab.txt\n",
      "2765236465.ko.tab.txt\n",
      "2765236465.ko.tab.txt\n",
      "2765236465.pfam.tab.txt\n",
      "2765236465.pfam.tab.txt\n",
      "2765236465.signalp.tab.txt\n",
      "2765236465.signalp.tab.txt\n",
      "2765236465.tigrfam.tab.txt\n",
      "2765236465.tigrfam.tab.txt\n",
      "2765236465.tmhmm.tab.txt\n",
      "2765236465.tmhmm.tab.txt\n",
      "README.txt\n",
      "2675903008.tar.gz\n",
      "2675903008.cog.tab.txt\n",
      "2675903008.fna\n",
      "2675903008.genes.faa\n",
      "2675903008.genes.fna\n",
      "2675903008.gff\n",
      "2675903008.intergenic.fna\n",
      "2675903008.ipr.tab.txt\n",
      "2675903008.ko.tab.txt\n",
      "2675903008.ko.tab.txt\n",
      "2675903008.pfam.tab.txt\n",
      "2675903008.pfam.tab.txt\n",
      "2675903008.signalp.tab.txt\n",
      "2675903008.signalp.tab.txt\n",
      "2675903008.tigrfam.tab.txt\n",
      "2675903008.tigrfam.tab.txt\n",
      "2675903008.tmhmm.tab.txt\n",
      "2675903008.tmhmm.tab.txt\n",
      "README.txt\n",
      "2695421014.tar.gz\n",
      "2695421014.cog.tab.txt\n",
      "2695421014.fna\n",
      "2695421014.genes.faa\n",
      "2695421014.gff\n",
      "2695421014.intergenic.fna\n",
      "2695421014.ipr.tab.txt\n",
      "2695421014.ko.tab.txt\n",
      "2695421014.ko.tab.txt\n",
      "2695421014.pfam.tab.txt\n",
      "2695421014.pfam.tab.txt\n",
      "2695421014.signalp.tab.txt\n",
      "2695421014.signalp.tab.txt\n",
      "2695421014.tigrfam.tab.txt\n",
      "2695421014.tigrfam.tab.txt\n",
      "2695421014.tmhmm.tab.txt\n",
      "2695421014.tmhmm.tab.txt\n",
      "README.txt\n",
      "2695421014.genes.FNA.wOutlier\n",
      "2695421014.genes.fna\n"
     ]
    }
   ],
   "source": [
    "sslc        = {'3211_1':'A','3211_3':'B','3211_5':'C','3211_6':'D','3212_2':'E','3212_3':'F','3212_4':'G','3212_5':'H','3213_3':'I','3214_6':'J'} # strain single leter code\n",
    "\n",
    "# Build the gene annotation dictionary\n",
    "annotations={'jgi_gene_id':[],'anno_type':[],'annos':[]}\n",
    "for strain in sslc.keys():\n",
    "    cur_strain = strain.replace('_','')\n",
    "    anno_dir = os.listdir('../annotations/Strsp{}_FD/IMG_Data'.format(cur_strain))\n",
    "    for af in anno_dir:\n",
    "        print(af)\n",
    "        if af.endswith('cog.tab.txt'):\n",
    "            with open('../annotations/Strsp{}_FD/IMG_Data/{}'.format(cur_strain,af)) as cog_file:\n",
    "                cog_df = pd.read_csv(cog_file,sep='\\t')\n",
    "        if af.endswith('.gff'):\n",
    "            with open('../annotations/Strsp{}_FD/IMG_Data/{}'.format(cur_strain,af)) as gff_file:\n",
    "                gff_df = pd.read_csv(gff_file,sep='\\t')\n",
    "        if af.endswith('ko.tab.txt'):\n",
    "            print(af)\n",
    "        if af.endswith('pfam.tab.txt'):\n",
    "            print(af)\n",
    "        if af.endswith('signalp.tab.txt'):\n",
    "            print(af)\n",
    "        if af.endswith('tigrfam.tab.txt'):\n",
    "            print(af)\n",
    "        if af.endswith('tmhmm.tab.txt'):\n",
    "            print(af)\n",
    "        #if af.endswith('crispr.txt'): # skipping this, since not all annotations have it\n",
    "        #    print(af)\n",
    "    \n",
    "# Build the community composition dictionary\n",
    "\n",
    "# Add the reads files using a separate tsv file, not the reads directory symbolic link\n",
    "\n",
    "libIDDict={}\n",
    "libSeqIDDict={}\n",
    "libCommType={}\n",
    "#readsFiles=os.listdir(\"../reads\") # Get read files names to build libSeqIDDict mapping\n",
    "with open(\"../RNAseq_library_names.tsv\",'r') as libraryFile:\n",
    "    reader = csv.reader(libraryFile,delimiter='\\t')\n",
    "    for line in reader:\n",
    "        if 'Library_Name' not in line[7]:\n",
    "            libName=line[7]\n",
    "            compositionList=line[10].replace('.','_')\n",
    "            seqID=line[12]\n",
    "            \n",
    "            \n",
    "            if ',' in compositionList:\n",
    "                compositionList=compositionList.split(',')\n",
    "            else:\n",
    "                compositionList=[compositionList]\n",
    "            libIDDict[libName]=compositionList\n",
    "            #curReadsFile=\"\"\n",
    "            #for f in readsFiles:\n",
    "            #    if seqID in f:\n",
    "            #        curReadsFile=f\n",
    "            #libSeqIDDict[libName]=curReadsFile\n",
    "            libCommType[libName]=line[-2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "dependent-damage",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_oid</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>percent_identity</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subj_start</th>\n",
       "      <th>subj_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>cog_id</th>\n",
       "      <th>cog_name</th>\n",
       "      <th>cog_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2698407776</td>\n",
       "      <td>180</td>\n",
       "      <td>21.32</td>\n",
       "      <td>38</td>\n",
       "      <td>172</td>\n",
       "      <td>2</td>\n",
       "      <td>137</td>\n",
       "      <td>6.000000e-08</td>\n",
       "      <td>50.8</td>\n",
       "      <td>COG0494</td>\n",
       "      <td>8-oxo-dGTP pyrophosphatase MutT and related ho...</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2698407779</td>\n",
       "      <td>133</td>\n",
       "      <td>19.82</td>\n",
       "      <td>9</td>\n",
       "      <td>107</td>\n",
       "      <td>10</td>\n",
       "      <td>120</td>\n",
       "      <td>4.000000e-04</td>\n",
       "      <td>37.3</td>\n",
       "      <td>COG2172</td>\n",
       "      <td>Anti-sigma regulatory factor (Ser/Thr protein ...</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2698407782</td>\n",
       "      <td>221</td>\n",
       "      <td>23.89</td>\n",
       "      <td>12</td>\n",
       "      <td>216</td>\n",
       "      <td>19</td>\n",
       "      <td>219</td>\n",
       "      <td>5.000000e-12</td>\n",
       "      <td>64.8</td>\n",
       "      <td>COG1434</td>\n",
       "      <td>Uncharacterized SAM-binding protein YcdF, DUF2...</td>\n",
       "      <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2698407791</td>\n",
       "      <td>428</td>\n",
       "      <td>26.57</td>\n",
       "      <td>1</td>\n",
       "      <td>398</td>\n",
       "      <td>1</td>\n",
       "      <td>390</td>\n",
       "      <td>1.000000e-37</td>\n",
       "      <td>151.0</td>\n",
       "      <td>COG0420</td>\n",
       "      <td>DNA repair exonuclease SbcCD nuclease subunit</td>\n",
       "      <td>390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2698407792</td>\n",
       "      <td>1060</td>\n",
       "      <td>20.68</td>\n",
       "      <td>1</td>\n",
       "      <td>724</td>\n",
       "      <td>1</td>\n",
       "      <td>703</td>\n",
       "      <td>1.000000e-33</td>\n",
       "      <td>139.0</td>\n",
       "      <td>COG0419</td>\n",
       "      <td>DNA repair exonuclease SbcCD ATPase subunit</td>\n",
       "      <td>908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4889</th>\n",
       "      <td>2698416243</td>\n",
       "      <td>161</td>\n",
       "      <td>27.56</td>\n",
       "      <td>16</td>\n",
       "      <td>142</td>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>2.000000e-07</td>\n",
       "      <td>48.8</td>\n",
       "      <td>COG1846</td>\n",
       "      <td>DNA-binding transcriptional regulator, MarR fa...</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4890</th>\n",
       "      <td>2698416246</td>\n",
       "      <td>256</td>\n",
       "      <td>34.81</td>\n",
       "      <td>117</td>\n",
       "      <td>250</td>\n",
       "      <td>1</td>\n",
       "      <td>132</td>\n",
       "      <td>2.000000e-24</td>\n",
       "      <td>106.0</td>\n",
       "      <td>COG2343</td>\n",
       "      <td>Uncharacterized conserved protein, DUF427 family</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4891</th>\n",
       "      <td>2698416248</td>\n",
       "      <td>421</td>\n",
       "      <td>27.43</td>\n",
       "      <td>1</td>\n",
       "      <td>275</td>\n",
       "      <td>1</td>\n",
       "      <td>278</td>\n",
       "      <td>4.000000e-16</td>\n",
       "      <td>79.7</td>\n",
       "      <td>COG4584</td>\n",
       "      <td>Transposase</td>\n",
       "      <td>278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4892</th>\n",
       "      <td>2698416249</td>\n",
       "      <td>286</td>\n",
       "      <td>42.32</td>\n",
       "      <td>41</td>\n",
       "      <td>279</td>\n",
       "      <td>17</td>\n",
       "      <td>254</td>\n",
       "      <td>3.000000e-56</td>\n",
       "      <td>212.0</td>\n",
       "      <td>COG1484</td>\n",
       "      <td>DNA replication protein DnaC</td>\n",
       "      <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4893</th>\n",
       "      <td>2698416251</td>\n",
       "      <td>237</td>\n",
       "      <td>31.74</td>\n",
       "      <td>24</td>\n",
       "      <td>187</td>\n",
       "      <td>12</td>\n",
       "      <td>177</td>\n",
       "      <td>7.000000e-16</td>\n",
       "      <td>77.8</td>\n",
       "      <td>COG1595</td>\n",
       "      <td>DNA-directed RNA polymerase specialized sigma ...</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4894 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_oid  gene_length  percent_identity  query_start  query_end  \\\n",
       "0     2698407776          180             21.32           38        172   \n",
       "1     2698407779          133             19.82            9        107   \n",
       "2     2698407782          221             23.89           12        216   \n",
       "3     2698407791          428             26.57            1        398   \n",
       "4     2698407792         1060             20.68            1        724   \n",
       "...          ...          ...               ...          ...        ...   \n",
       "4889  2698416243          161             27.56           16        142   \n",
       "4890  2698416246          256             34.81          117        250   \n",
       "4891  2698416248          421             27.43            1        275   \n",
       "4892  2698416249          286             42.32           41        279   \n",
       "4893  2698416251          237             31.74           24        187   \n",
       "\n",
       "      subj_start  subj_end        evalue  bit_score   cog_id  \\\n",
       "0              2       137  6.000000e-08       50.8  COG0494   \n",
       "1             10       120  4.000000e-04       37.3  COG2172   \n",
       "2             19       219  5.000000e-12       64.8  COG1434   \n",
       "3              1       390  1.000000e-37      151.0  COG0420   \n",
       "4              1       703  1.000000e-33      139.0  COG0419   \n",
       "...          ...       ...           ...        ...      ...   \n",
       "4889           1       122  2.000000e-07       48.8  COG1846   \n",
       "4890           1       132  2.000000e-24      106.0  COG2343   \n",
       "4891           1       278  4.000000e-16       79.7  COG4584   \n",
       "4892          17       254  3.000000e-56      212.0  COG1484   \n",
       "4893          12       177  7.000000e-16       77.8  COG1595   \n",
       "\n",
       "                                               cog_name  cog_length  \n",
       "0     8-oxo-dGTP pyrophosphatase MutT and related ho...         161  \n",
       "1     Anti-sigma regulatory factor (Ser/Thr protein ...         146  \n",
       "2     Uncharacterized SAM-binding protein YcdF, DUF2...         223  \n",
       "3         DNA repair exonuclease SbcCD nuclease subunit         390  \n",
       "4           DNA repair exonuclease SbcCD ATPase subunit         908  \n",
       "...                                                 ...         ...  \n",
       "4889  DNA-binding transcriptional regulator, MarR fa...         126  \n",
       "4890   Uncharacterized conserved protein, DUF427 family         132  \n",
       "4891                                        Transposase         278  \n",
       "4892                       DNA replication protein DnaC         254  \n",
       "4893  DNA-directed RNA polymerase specialized sigma ...         182  \n",
       "\n",
       "[4894 rows x 12 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "green-creator",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>##gff-version 3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">Ga0123419_11</th>\n",
       "      <th rowspan=\"11\" valign=\"top\">img_core_v400</th>\n",
       "      <th rowspan=\"11\" valign=\"top\">CDS</th>\n",
       "      <th>1</th>\n",
       "      <th>2544</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>0</th>\n",
       "      <td>ID=2698407768;locus_tag=Ga0123419_0001;product...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2732</th>\n",
       "      <th>3391</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>0</th>\n",
       "      <td>ID=2698407769;locus_tag=Ga0123419_0002;product...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3528</th>\n",
       "      <th>4229</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>0</th>\n",
       "      <td>ID=2698407770;locus_tag=Ga0123419_0003;product...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4226</th>\n",
       "      <th>4576</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>0</th>\n",
       "      <td>ID=2698407771;locus_tag=Ga0123419_0004;product...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4861</th>\n",
       "      <th>5886</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>0</th>\n",
       "      <td>ID=2698407772;locus_tag=Ga0123419_0005;product...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9372836</th>\n",
       "      <th>9373078</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>0</th>\n",
       "      <td>ID=2698416254;locus_tag=Ga0123419_8487;product...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9373320</th>\n",
       "      <th>9373534</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>0</th>\n",
       "      <td>ID=2698416255;locus_tag=Ga0123419_8488;product...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9373676</th>\n",
       "      <th>9374488</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>0</th>\n",
       "      <td>ID=2698416256;locus_tag=Ga0123419_8489;product...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9374906</th>\n",
       "      <th>9375286</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>0</th>\n",
       "      <td>ID=2698416257;locus_tag=Ga0123419_8490;product...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9375339</th>\n",
       "      <th>9375451</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>0</th>\n",
       "      <td>ID=2698416258;locus_tag=Ga0123419_8491;product...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8491 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                        ##gff-version 3\n",
       "Ga0123419_11 img_core_v400 CDS 1       2544    . + 0  ID=2698407768;locus_tag=Ga0123419_0001;product...\n",
       "                               2732    3391    . - 0  ID=2698407769;locus_tag=Ga0123419_0002;product...\n",
       "                               3528    4229    . - 0  ID=2698407770;locus_tag=Ga0123419_0003;product...\n",
       "                               4226    4576    . - 0  ID=2698407771;locus_tag=Ga0123419_0004;product...\n",
       "                               4861    5886    . - 0  ID=2698407772;locus_tag=Ga0123419_0005;product...\n",
       "...                                                                                                 ...\n",
       "                               9372836 9373078 . + 0  ID=2698416254;locus_tag=Ga0123419_8487;product...\n",
       "                               9373320 9373534 . + 0  ID=2698416255;locus_tag=Ga0123419_8488;product...\n",
       "                               9373676 9374488 . + 0  ID=2698416256;locus_tag=Ga0123419_8489;product...\n",
       "                               9374906 9375286 . + 0  ID=2698416257;locus_tag=Ga0123419_8490;product...\n",
       "                               9375339 9375451 . + 0  ID=2698416258;locus_tag=Ga0123419_8491;product...\n",
       "\n",
       "[8491 rows x 1 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "resistant-richards",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BCUTC_ten-member_ABCDEFGHIJ.genes.fna_toss_counts.txt\n"
     ]
    }
   ],
   "source": [
    "def get_jgi_geneid(x):\n",
    "    return x.split(' ')[0]\n",
    "def get_assembly_geneid(x):\n",
    "    return x.split(' ')[1]\n",
    "\n",
    "counts_files=os.listdir('../counts')\n",
    "for ctr,cf in enumerate(counts_files):\n",
    "    if cf.endswith('counts.txt') and ctr ==0:\n",
    "        print(cf)\n",
    "        cf_header = cf.split('.')[0]\n",
    "        lib, complexity, ref = cf_header.split('_')\n",
    "        cur_df = pd.read_csv('../counts/{}'.format(cf),sep='\\t',comment='#',header=None)\n",
    "        cur_df.columns = ['GeneID','Length','Bases','Coverage','Reads','RPKM','Frags','FPKM']\n",
    "        # Add the ref used\n",
    "        cur_df.loc[:,'reference'] = ref\n",
    "        # Add the library name\n",
    "        cur_df.loc[:,'library'] = lib\n",
    "        # Add the complexity\n",
    "        cur_df.loc[:,'complexity'] = complexity\n",
    "        # Add the community composition\n",
    "        \n",
    "        # Add JGI gene id\n",
    "        cur_df.loc[:,'jgi_gene_id'] = cur_df.apply(lambda x: get_jgi_geneid(x.GeneID),axis=1)\n",
    "        \n",
    "        # Add JGI assembly gene id\n",
    "        cur_df.loc[:,'assembly_gene_id'] = cur_df.apply(lambda x: get_assembly_geneid(x.GeneID),axis=1)\n",
    "        \n",
    "        # Add the TPM normalization\n",
    "        cur_df=cur_df.groupby(['library']).apply(lambda x: x.assign(tpm=lambda x: (x.Reads  / (x.Length/1e3)) / ((x.Reads  / (x.Length/1e3)).sum() / 1e6 )  )).droplevel(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "determined-battery",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Length</th>\n",
       "      <th>Bases</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Reads</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>Frags</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>reference</th>\n",
       "      <th>library</th>\n",
       "      <th>complexity</th>\n",
       "      <th>jgi_gene_id</th>\n",
       "      <th>assembly_gene_id</th>\n",
       "      <th>tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2785552622 Ga0151165_0001 1..139(-)(Ga0151165_...</td>\n",
       "      <td>139</td>\n",
       "      <td>19851</td>\n",
       "      <td>142.8129</td>\n",
       "      <td>192</td>\n",
       "      <td>405.1146</td>\n",
       "      <td>96</td>\n",
       "      <td>405.1146</td>\n",
       "      <td>ABCDEFGHIJ</td>\n",
       "      <td>BCUTC</td>\n",
       "      <td>ten-member</td>\n",
       "      <td>2785552622</td>\n",
       "      <td>Ga0151165_0001</td>\n",
       "      <td>227.852749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2785552623 Ga0151165_0002 548..1774(-)(Ga01511...</td>\n",
       "      <td>1227</td>\n",
       "      <td>300</td>\n",
       "      <td>0.2445</td>\n",
       "      <td>2</td>\n",
       "      <td>0.4781</td>\n",
       "      <td>1</td>\n",
       "      <td>0.4781</td>\n",
       "      <td>ABCDEFGHIJ</td>\n",
       "      <td>BCUTC</td>\n",
       "      <td>ten-member</td>\n",
       "      <td>2785552623</td>\n",
       "      <td>Ga0151165_0002</td>\n",
       "      <td>0.268877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2785552624 Ga0151165_0003 1776..2504(-)(Ga0151...</td>\n",
       "      <td>729</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>ABCDEFGHIJ</td>\n",
       "      <td>BCUTC</td>\n",
       "      <td>ten-member</td>\n",
       "      <td>2785552624</td>\n",
       "      <td>Ga0151165_0003</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2785552625 Ga0151165_0004 2562..4184(-)(Ga0151...</td>\n",
       "      <td>1623</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>ABCDEFGHIJ</td>\n",
       "      <td>BCUTC</td>\n",
       "      <td>ten-member</td>\n",
       "      <td>2785552625</td>\n",
       "      <td>Ga0151165_0004</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2785552626 Ga0151165_0005 5197..7863(+)(Ga0151...</td>\n",
       "      <td>2667</td>\n",
       "      <td>300</td>\n",
       "      <td>0.1125</td>\n",
       "      <td>2</td>\n",
       "      <td>0.2199</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2199</td>\n",
       "      <td>ABCDEFGHIJ</td>\n",
       "      <td>BCUTC</td>\n",
       "      <td>ten-member</td>\n",
       "      <td>2785552626</td>\n",
       "      <td>Ga0151165_0005</td>\n",
       "      <td>0.123701</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              GeneID  Length  Bases  Coverage  \\\n",
       "0  2785552622 Ga0151165_0001 1..139(-)(Ga0151165_...     139  19851  142.8129   \n",
       "1  2785552623 Ga0151165_0002 548..1774(-)(Ga01511...    1227    300    0.2445   \n",
       "2  2785552624 Ga0151165_0003 1776..2504(-)(Ga0151...     729      0    0.0000   \n",
       "3  2785552625 Ga0151165_0004 2562..4184(-)(Ga0151...    1623      0    0.0000   \n",
       "4  2785552626 Ga0151165_0005 5197..7863(+)(Ga0151...    2667    300    0.1125   \n",
       "\n",
       "   Reads      RPKM  Frags      FPKM   reference library  complexity  \\\n",
       "0    192  405.1146     96  405.1146  ABCDEFGHIJ   BCUTC  ten-member   \n",
       "1      2    0.4781      1    0.4781  ABCDEFGHIJ   BCUTC  ten-member   \n",
       "2      0    0.0000      0    0.0000  ABCDEFGHIJ   BCUTC  ten-member   \n",
       "3      0    0.0000      0    0.0000  ABCDEFGHIJ   BCUTC  ten-member   \n",
       "4      2    0.2199      1    0.2199  ABCDEFGHIJ   BCUTC  ten-member   \n",
       "\n",
       "  jgi_gene_id assembly_gene_id         tpm  \n",
       "0  2785552622   Ga0151165_0001  227.852749  \n",
       "1  2785552623   Ga0151165_0002    0.268877  \n",
       "2  2785552624   Ga0151165_0003    0.000000  \n",
       "3  2785552625   Ga0151165_0004    0.000000  \n",
       "4  2785552626   Ga0151165_0005    0.123701  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cur_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "banned-ranking",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3 (biotools)",
   "language": "python",
   "name": "biotools"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
